src/snowflake/snowpark/_internal/udf_utils.py

```diff
@@ -381,6 +384,76 @@ def get_types_from_type_hints(
     return return_type, input_types
 
 
+def get_opt_arg_defaults(
+    func: Union[Callable, Tuple[str, str]],
+    object_type: TempObjectType,
+    input_types: List[DataType],
+) -> List[Optional[str]]:
+    EMPTY_DEFAULT_VALUES = [None] * len(input_types)
+
+    def build_default_values_result(
+        default_values: Any, input_types: List[DataType]
+    ) -> List[Optional[str]]:
+        num_optional_args = len(default_values)
+        num_positional_args = len(input_types) - num_optional_args
+        default_values_to_sql_str = [
+            to_sql(value, datatype)
+            for value, datatype in zip(default_values, input_types[-num_optional_args:])
+        ]
+        return [None] * num_positional_args + default_values_to_sql_str
+
+    def get_opt_arg_defaults_from_callable():
+        target_func = None
+        if object_type == TempObjectType.TABLE_FUNCTION:
+            # extract from process method
+            if hasattr(func, TABLE_FUNCTION_PROCESS_METHOD):
+                target_func = getattr(func, TABLE_FUNCTION_PROCESS_METHOD)
+        if object_type in (TempObjectType.PROCEDURE, TempObjectType.FUNCTION):
+            # sproc and udf
+            target_func = func
+
+        if target_func is None:
+            return EMPTY_DEFAULT_VALUES
+
+        arg_spec = inspect.getfullargspec(target_func)
+        return build_default_values_result(arg_spec.defaults, input_types)
+
+    def get_opt_arg_defaults_from_file():
+        filename, func_name = func[0], func[1]
+        default_values_str = None
+        if not is_local_python_file(filename):
+            return EMPTY_DEFAULT_VALUES
+
+        if object_type == TempObjectType.TABLE_FUNCTION:
+            default_values_str = retrieve_func_defaults_from_source(
+                filename, TABLE_FUNCTION_PROCESS_METHOD, func_name
```

Comment:
just curious: the `func_time` in this case is actually the table function class name?

Reply:
you mean `func_name`?